### PR TITLE
feat: migrate JWT signing queries

### PR DESCRIPTION
**Pull request by anugrahsinghal**
- [x] create DO for JWTSigning
- [x] migrate queries
- [x] add exception handling for HibernateSpecificException if thrown
- [x] all tests should pass

### CODE REVIEW

**Comment by anugrahsinghal** [comment]
@rishabhpoddar 

https://github.com/supertokens/supertokens-sql-plugin/pull/37/files#diff-5f72db7c99e6138c265f04d0edf32fd2dde8bc402f7de493e60668e22ddc0666R980-R983


what are your opinions about this ?

File: src/main/java/io/supertokens/storage/sql/Start.java [hunk 1]
```diff
@@ -972,15 +973,28 @@ public List<JWTSigningKeyInfo> getJWTSigningKeys_Transaction(TransactionConnecti
...
-            if (e instanceof PSQLException && isPrimaryKeyError(((PSQLException) e).getServerErrorMessage(),
-                    Config.getConfig(this).getJWTSigningKeysTable())) {
+            JWTSigningQueries.setJWTSigningKeyInfo_Transaction(session, info);
+        } catch (PersistenceException eTemp) {
+            // when trying to save a Object within same hibernate session
+            // it throws a NonUniqueObjectException
+            // TODO: sql-plugin -> does this really seem like a real scenario
+            // session is not shared between 2 consecutive calls
+            if(eTemp instanceof NonUniqueObjectException) {
                 throw new DuplicateKeyIdException();
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/sql/Start.java]: To discuss this on call

File: src/main/java/io/supertokens/storage/sql/Start.java [hunk 2]
```diff
@@ -972,15 +973,28 @@ public List<JWTSigningKeyInfo> getJWTSigningKeys_Transaction(TransactionConnecti
...
+            PostgreSQLConfig config = Config.getConfig(this);
+            ServerErrorMessage serverMessage = psqlException.getServerErrorMessage();
+
+           if (isPrimaryKeyError(serverMessage, config.getJWTSigningKeysTable())) {
+               throw new DuplicateKeyIdException();
+            }
+
+            throw new StorageQueryException(eTemp);
+        } catch (SQLException e) {
             // We keep the old exception detection logic to ensure backwards compatibility.
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/sql/Start.java:1013]: You can remove the old way of doing exception handling

File: src/main/java/io/supertokens/storage/sql/domainobject/jwtsigning/JWTSigningDO.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+@Table(name = "jwt_signing_keys")
+public class JWTSigningDO extends PrimaryKeyFetchable {
+
+    @Id
+    @Column(nullable = false)
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/sql/domainobject/jwtsigning/JWTSigningDO.java]: Specify length

File: src/main/java/io/supertokens/storage/sql/domainobject/jwtsigning/JWTSigningDO.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    private String key_id;
+
+    @Column(columnDefinition = "TEXT")
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/sql/domainobject/jwtsigning/JWTSigningDO.java]: nullable false

File: src/main/java/io/supertokens/storage/sql/domainobject/jwtsigning/JWTSigningDO.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.io.Serializable;
+
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/sql/domainobject/jwtsigning/JWTSigningDO.java:30]: Add comment of raw SQL table schema

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/sql/Start.java (hunk 1 above)
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/sql/Start.java]: We can try doing a flush inside `setJWTSigningKeyInfo_Transaction` and that may solve this issue where you don't need to do `if(eTemp instanceof NonUniqueObjectException) {`

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Comment by anugrahsinghal** [comment]
https://github.com/supertokens/supertokens-sql-plugin/pull/37#discussion_r926789690

For this issue, a `flush` will not be able to help
- The issue originates since we are trying to `save` an entity with `same Id` within the `same` transaction
And since flush does not clear the hibernate cache, this exception is being thrown by hibernate itself

---
Possible Solutions
- Change `ExceptionParsingTest::setJWTSigningKey_TransactionExceptions` to use 2 transactions instead of one to test this scenario, this will make the validation go to the DB 
- Keep this exception handling as-is for now and handle it when we start with designing/improving the exception handling part later on


@rishabhpoddar - WDYT?

**Comment by rishabhpoddar** [comment]
@anugrahsinghal i think we should keep the test as it is and deal with the exception handling later on (resolved the associated review comment).

File: src/main/java/io/supertokens/storage/sql/Start.java (hunk 1 above)
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/sql/Start.java]: https://github.com/supertokens/supertokens-sql-plugin/pull/37#issuecomment-1199052419

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**